addons/sale_expense/tests/test_reinvoice.py

```diff
@@ -9,109 +9,227 @@
+        cls.company_data.update({
+            'service_order_sales_price': cls.env['product.product'].with_company(cls.company_data['company']).create({
+                'name': 'service_order_sales_price',
+                'categ_id': cls.company_data['product_category'].id,
+                'standard_price': 235.0,
+                'list_price': 280.0,
+                'type': 'service',
+                'weight': 0.01,
+                'uom_id': cls.env.ref('uom.product_uom_unit').id,
+                'uom_po_id': cls.env.ref('uom.product_uom_unit').id,
+                'default_code': 'FURN_9999',
+                'invoice_policy': 'order',
+                'expense_policy': 'sales_price',
+                'taxes_id': [Command.set([])],
+                'supplier_taxes_id': [Command.set([])],
+                'can_be_expensed': True,
+            }),
+            'service_delivery_sales_price': cls.env['product.product'].with_company(cls.company_data['company']).create({
+                'name': 'service_order_sales_price',
+                'categ_id': cls.company_data['product_category'].id,
+                'standard_price': 235.0,
+                'list_price': 280.0,
+                'type': 'service',
+                'weight': 0.01,
+                'uom_id': cls.env.ref('uom.product_uom_unit').id,
+                'uom_po_id': cls.env.ref('uom.product_uom_unit').id,
+                'default_code': 'FURN_9999',
+                'invoice_policy': 'order',
+                'expense_policy': 'sales_price',
+                'taxes_id': [Command.set([])],
+                'supplier_taxes_id': [Command.set([])],
+                'can_be_expensed': True,
+            }),
         })
 
+    def test_expenses_reinvoice(self):
+        """
+        Test that expenses are re-invoiced correctly and that the quantity is updated when it has to.
+            - Lines are never grouped together (even if reinvoced at sale price and with a re-invoice delivered policy)
+            - When posting an expense, it creates the corresponding sol with the expense quantity
+            - The quantities ordered and delivered are reset to 0 when:
+                - the expense sheet is unposted
+                - the expense move is reversed
+                - the expense move is reset to draft
+        """
         # create SO line and confirm SO (with only one line)
         sale_order = self.env['sale.order'].with_context(mail_notrack=True, mail_create_nolog=True).create({
             'partner_id': self.partner_a.id,
             'partner_invoice_id': self.partner_a.id,
             'partner_shipping_id': self.partner_a.id,
-            'order_line': [(0, 0, {
-                'name': self.company_data['product_order_sales_price'].name,
-                'product_id': self.company_data['product_order_sales_price'].id,
-                'product_uom_qty': 2.0,
-                'price_unit': 1000.0,
-            })],
+            'order_line': [],
```

Comment:
it's weird to have a confirmed order with no lines (especially with the comment above)
maybe add a line that could be confused for the expense line

Reply:
Indeed, too eager to clean useless stuff :p